### PR TITLE
refactor(AddressRange): create and use a proper type

### PR DESCRIPTION
**Pull request by n1tram1**
Previously AddressRange was just a type alias to a usize pair. This caused confusion as sometimes it was (addr, size) or (start, end) and lead to a few confusing debugs.

The new type hopefully clear up confusions and make it easy to iterate over the pages in an addressrange.

### CODE REVIEW

File: hal_core/src/lib.rs [hunk 1]
```diff
@@ -7,27 +7,53 @@ pub enum Error {}
...
+    }
+
+    pub fn with_size(start: usize, size: usize) -> Self {
+        Self::new(start, start + size)
+    }
+
+    pub fn round_up_to_page(self, page_size: usize) -> Self {
+        Self {
+            start: self.start,
+            end: mm::align_up(self.end, page_size),
```
Comment (Skallwar) [inline hal_core/src/lib.rs:38]: Could be cool if `mm` was able to find the page_size by itself

File: hal_core/src/lib.rs [hunk 2]
```diff
@@ -7,27 +7,53 @@ pub enum Error {}
...
+        }
+    }
+
+    pub fn iter_pages(self, page_size: usize) -> impl Iterator<Item = usize> {
```
Comment (Skallwar) [inline hal_core/src/lib.rs:42]: I wonder if we should just provide `page_size` to new. Are we using `AddressRange` for something else than pages?

File: hal_core/src/lib.rs [hunk 3]
```diff
@@ -7,27 +7,53 @@ pub enum Error {}
...
+        assert_eq!(self.end, mm::align_up(self.end, page_size));
+
+        (self.start..=self.end).step_by(page_size)
+    }
+
+    pub fn count_pages(&self, page_size: usize) -> usize {
+        mm::align_up(self.size(), page_size) / page_size
+    }
```
Comment (Skallwar) [inline hal_core/src/lib.rs:50]: Dito

File: hal_riscv64/src/mm/mod.rs [hunk 4]
```diff
@@ -50,3 +50,7 @@ unsafe fn load_pagetable(pt: &'static mut PageTable) {
         asm!("sfence.vma");
     }
 }
+
+pub fn align_up(addr: usize) -> usize {
+   mm::align_up(addr, PageTable::PAGE_SIZE)
+}
```
Comment (Skallwar) [inline hal_riscv64/src/mm/mod.rs:56]: I think in the future we should just merge hal_core with hal_archXXX with conditional compilation. There is place for some code factoring but that for another time

**Comment by n1tram1** [comment]
I agree with all the remarks about `page_size`.
This is an issue I want to address but I would rather do it outside of this PR if ok ?

File: hal_core/src/lib.rs (hunk 2 above)
Comment (Skallwar) [reply hal_core/src/lib.rs:39]: We don't use it for anything else than pages, I've checked

File: hal_core/src/lib.rs [hunk 5]
```diff
@@ -7,27 +7,50 @@ pub enum Error {}
...
+/// A range similar to core::ops::Range but that is copyable.
+/// The range is half-open, inclusive below, exclusive above, ie. [start; end[
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct AddressRange {
+    pub start: usize,
+    pub end: usize,
+}
+
+impl AddressRange {
+    pub fn new(start: usize, end: usize) -> Self {
```
Comment (CohenArthur) [inline hal_core/src/lib.rs:19]: Could be nice to have a constructor from a `Range<usize>`, the syntax would be a bit more range like: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=38449a783dbfc3192aba75f8c9f6de67

Not sure if that's useful though!

**Review by CohenArthur — APPROVED** [review]

**Review by Skallwar — APPROVED** [review]

**Pull request merged**